### PR TITLE
Add metadata with the Reboot API in order to add offsets onto coordinate data

### PR DESCRIPTION
**Pull request by jimmybuffi**
## Description

Accordingly, this PR implements the Reboot Motion API in order to use the data export API to download the metadata. And then a function is added to the toolkit backend to add the offsets back onto coordinate data.

The reboot motion api implementation can be found in `reboot_toolkit/reboot_motion_api.py` and it was inspired by the way Stripe organizes their python client here:
https://github.com/stripe/stripe-python/blob/master/stripe/_stripe_client.py#L103

## Motivation and Context

Because Reboot Motion removes offsets from coordinate systems for certain motion capture types, Users need to be able to add these offsets back onto the coordinate data when they run analyses. These offsets can be found in Reboot Motion's stored metadata.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Enhancement (enhances currently existing functionality or changes it in some way that is non-breaking and also not addressing a bug)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] My code requires a new environment variable or other globally accessible value.

### CODE REVIEW

File: reboot_toolkit/toolkit_backend.py
Comment (jimmybuffi) [inline reboot_toolkit/toolkit_backend.py:1]: Sorry for all the changes here, lots of these are just formatting changes using `black`. The impactful changes are the new functions are at the bottom: 

export_data
add_offsets_from_metadata

File: reboot_toolkit/reboot_motion_api.py
Comment (jimmybuffi) [inline reboot_toolkit/reboot_motion_api.py:1]: At some point this could be moved to its own repository and installed as a dependency, so I tried to give this file as few external dependencies as possible

File: reboot_toolkit/reboot_motion_api.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+
+    def __exit__(self, exc_type, exc_val, exc_traceback):
+        """Close the requests session when exiting a context manager."""
+        self.close()
+
+    def _request(
+        self,
+        method: str,
+        route: str,
+        params: dict | list | None = None,
```
Comment (RobertCalise) [inline reboot_toolkit/reboot_motion_api.py:94]: From what I see in the `requests` library, I think this type needs to be list[tuple] - supports lists only as lists of tuples when passed in as params (i.e.: `[('param1', 0), ('param2', 1)]`).

File: reboot_toolkit/reboot_motion_api.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+from concurrent.futures import ThreadPoolExecutor
+from io import BytesIO
+from itertools import repeat
+
+import pyarrow as pa
+import requests
+import warnings
+
+
+def locals_to_input(local_vars: dict) -> dict:
```
Comment (RobertCalise) [inline reboot_toolkit/reboot_motion_api.py:26]: When locals() is passed into this function from inside a class, I think there might be some additional keys that need to be filtered (like `__module__`) - maybe we want to filter any key that starts with `__`?

**Review by RobertCalise — COMMENTED** [review]
Submitting comments as-is, since @jimmybuffi and I discussed this in person; will re-review after some additional changes.

File: reboot_toolkit/reboot_motion_api.py (hunk 2 above)
Comment (jimmybuffi) [reply reboot_toolkit/reboot_motion_api.py:26]: do you have the link for where you saw that this would be the case? 

i've been using this locals() function just fine and i haven't seen any keys like this, but i'm curious to see the resource that mentioned this as a possibility

File: reboot_toolkit/reboot_motion_api.py (hunk 1 above)
Comment (jimmybuffi) [reply reboot_toolkit/reboot_motion_api.py:94]: yeah we can make these types more specific at some point for sure

File: reboot_toolkit/reboot_motion_api.py (hunk 2 above)
Comment (RobertCalise) [reply reboot_toolkit/reboot_motion_api.py:26]: @jimmybuffi I'm going to resolve this since we discussed on Slack, no need to modify anything here.

File: reboot_toolkit/toolkit_backend.py [hunk 3]
```diff
@@ -862,42 +1031,103 @@ def save_figs_to_html(
...
+    if movement_type_enum.value not in accepted_movement_types:
+        raise NotImplementedError(
+            "Adding offsets only implemented for movement types: {}, others coming soon!".format(
+                accepted_movement_types
+            )
+        )
+
+    assert set(data_df["org_movement_id"]) == set(
+        metadata_df["org_movement_id"]
+    ), "Mismatched org movement ids in data_df and metadata_df, but must be matched to add offsets"
```
Comment (cpjansen) [inline reboot_toolkit/toolkit_backend.py]: I'm wondering if we might need to have a bit more detail in the error statement or maybe reword it. I'm not sure if saying the org movement ids are "mismatched" will be immediately clear to the user on what they did incorrectly (maybe I'm underestimating our users).
Maybe adding an additional sentence along these lines: "The org movement ids present in metadata_df must match exactly the org movement ids in data_df"

Another possibility could be to include an example in a demo where metadata_df is missing plays from data_df so they have an example of how they could handle a mismatch

File: reboot_toolkit/toolkit_backend.py [hunk 4]
```diff
@@ -862,42 +1031,103 @@ def save_figs_to_html(
...
+
+    if movement_type_enum.value not in accepted_movement_types:
+        raise NotImplementedError(
+            "Adding offsets only implemented for movement types: {}, others coming soon!".format(
+                accepted_movement_types
+            )
+        )
+
+    assert set(data_df["org_movement_id"]) == set(
+        metadata_df["org_movement_id"]
```
Comment (cpjansen) [inline reboot_toolkit/toolkit_backend.py]: Would it cause issues if there were extra plays in metadata_df that aren't in data_df? I was thinking that would actually be ok as long as we have metadata for every play in data_df
If it is ok, then rather than checking for exact match, potentially using logic along these lines to only do a check for plays that are in data_df but not metadata_df.
`assert len(set(data_df["org_movement_id"]).difference(set(metadata_df["org_movement_id"]))) == 0`

File: reboot_toolkit/toolkit_backend.py [hunk 5]
```diff
@@ -862,42 +1031,103 @@ def save_figs_to_html(
...
+    offset_cols = [
+        "org_movement_id",
+        "X_offset",
+        "Y_offset",
+        "Z_offset",
+    ]
+
+    data_df = data_df.merge(metadata_df[offset_cols], on="org_movement_id", how="left")
+
+    cols_to_drop = []
```
Comment (cpjansen) [inline reboot_toolkit/toolkit_backend.py]: The values for `cols_to_drop` are listed out in `offset_cols` along with org_movement_id, but then are being created by appending to a list here. It feels somewhat redundant.

File: reboot_toolkit/toolkit_backend.py (hunk 3 above)
Comment (jimmybuffi) [reply reboot_toolkit/toolkit_backend.py]: great point will update this message

File: reboot_toolkit/toolkit_backend.py (hunk 4 above)
Comment (jimmybuffi) [reply reboot_toolkit/toolkit_backend.py]: yeah could def implement this, i was thinking to be really strict with what we allow through, but you're right that it's ok if there are more IDs int he metadata than in the data

File: reboot_toolkit/toolkit_backend.py (hunk 5 above)
Comment (jimmybuffi) [reply reboot_toolkit/toolkit_backend.py]: yeah i guess i did this as a way to exclude the org movement ID from the columns to drop, but could do this a little differently so it's not as redundant

**Review by cpjansen — APPROVED** [review]

File: reboot_toolkit/toolkit_backend.py [hunk 6]
```diff
@@ -862,42 +1031,104 @@ def save_figs_to_html(
...
+    accepted_movement_types = {"baseball-hitting"}
+
+    if movement_type_enum.value not in accepted_movement_types:
+        raise NotImplementedError(
+            "Adding offsets only implemented for movement types: {}, others coming soon!".format(
+                accepted_movement_types
+            )
+        )
+
+    assert set(data_df["org_movement_id"]).issubset(
```
Comment (RobertCalise) [inline reboot_toolkit/toolkit_backend.py]: We should replace the assert statement here with a regular conditional that raises an exception.

File: reboot_toolkit/toolkit_backend.py [hunk 7]
```diff
@@ -584,7 +754,9 @@ def filter_pop_df(
     :return: the filtered and down-sampled population dataframe
     """
 
-    assert len(mean_df) == len(std_df), "Mismatched frames between mean and std population data"
+    assert len(mean_df) == len(
```
Comment (RobertCalise) [inline reboot_toolkit/toolkit_backend.py]: We should replace the assert statement here with a regular conditional that raises an exception.

**Review by RobertCalise — APPROVED** [review]
Other than a couple of minor modifications, this looks good to me!
